internal/query/roles_v2.go

```diff
@@ -0,0 +1,668 @@
+	if roleResourceType == nil {
+		return nil
+	}
+
+	roleRef := resourceToSpiceDBRef(e.namespace, roleResource)
+	ownerRef := resourceToSpiceDBRef(e.namespace, owner)
+
+	// e.g., rolev2:super-admin#owner@tenant:tnntten-root
+	ownerRel := &pb.RelationshipUpdate{
+		Operation: pb.RelationshipUpdate_OPERATION_TOUCH,
+		Relationship: &pb.Relationship{
+			Resource: roleRef,
+			Relation: iapl.RoleOwnerRelation,
+			Subject: &pb.SubjectReference{
+				Object: ownerRef,
+			},
+		},
+	}
+
+	// e.g., tenant:tnntten-root#member_role@rolev:super-admin
+	memberRel := &pb.RelationshipUpdate{
+		Operation: pb.RelationshipUpdate_OPERATION_TOUCH,
+		Relationship: &pb.Relationship{
+			Resource: ownerRef,
+			Relation: iapl.RoleOwnerMemberRoleRelation,
+			Subject: &pb.SubjectReference{
+				Object: roleRef,
+			},
+		},
+	}
+
+	return []*pb.RelationshipUpdate{ownerRel, memberRel}
+}
+
+// createRoleV2RelationshipUpdatesForAction creates permission relationship lines in role
+// i.e., role:<role_name>#<action>_rel@<namespace>/<subjType>:*
+func (e *engine) createRoleV2RelationshipUpdatesForAction(
+	action string,
+	roleRef *pb.ObjectReference,
+	op pb.RelationshipUpdate_Operation,
+) []*pb.RelationshipUpdate {
+	rels := make([]*pb.RelationshipUpdate, len(e.rbac.RoleSubjectTypes))
+
+	for i, subjType := range e.rbac.RoleSubjectTypes {
+		rels[i] = &pb.RelationshipUpdate{
+			Operation: op,
+			Relationship: &pb.Relationship{
+				Resource: roleRef,
+				Relation: actionToRelation(action),
+				Subject: &pb.SubjectReference{
+					Object: &pb.ObjectReference{
+						ObjectType: e.namespaced(subjType),
+						ObjectId:   "*",
+					},
+				},
+			},
+		}
+	}
+
+	return rels
+}
+
+// roleV2Relationships creates relationships between a V2 role and its permissions.
+func (e *engine) roleV2Relationships(role types.Role) []*pb.RelationshipUpdate {
+	var rels []*pb.RelationshipUpdate
+
+	roleResource, err := e.NewResourceFromID(role.ID)
+	if err != nil {
+		panic(err)
+	}
+
+	roleResourceType := e.GetResourceType(e.rbac.RoleResource.Name)
+	if roleResourceType == nil {
+		return rels
+	}
+
+	roleRef := resourceToSpiceDBRef(e.namespace, roleResource)
+
+	for _, action := range role.Actions {
+		rels = append(
+			rels,
+			e.createRoleV2RelationshipUpdatesForAction(
+				action, roleRef,
+				pb.RelationshipUpdate_OPERATION_TOUCH,
+			)...,
+		)
+	}
+
+	return rels
+}
+
+func (e *engine) listRoleV2Actions(ctx context.Context, role types.Role) ([]string, error) {
+	if len(e.rbac.RoleSubjectTypes) == 0 {
+		return nil, nil
+	}
+
+	// there could be multiple subjects for a permission,
+	// e.g.
+	//   infratographer/rolev2:lb_viewer#loadbalancer_get_rel@infratographer/user:*
+	//   infratographer/rolev2:lb_viewer#loadbalancer_get_rel@infratographer/client:*
+	// here we only need one of them since the action is the only thing we care
+	// about
+	permRelationshipSubjType := e.namespaced(e.rbac.RoleSubjectTypes[0])
```

Comment:
I wonder if there's actually a better way of doing this. We could, for example, define a resource called `subject_set` or something and add a bunch of relationships to `user:*`, `client:*`, and so on as needed.

Maybe not work for this PR, though.